File: backend/app/core/numeric.py
"""Decimal arithmetic for money and quantities.

The `NUMERIC -> JavaScript number` mapping is the only lossy conversion in the
schema, so the rounding contract is implemented in one place rather than
re-derived at each call site (Data_Requirements_Database.md section 7.1).

The rule: both sides round identically at the boundary. The client rounds
before writing, the server rounds before storing, and all server-side
arithmetic uses `Decimal` rather than `float`. Because the stored values are
then bit-identical on device and server, last-write-wins never fires on a
value that merely *looks* different.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal, InvalidOperation

QUANTITY_EXP = Decimal("0.001")   # 3 dp: 0.250 kg, 12.500 L, 1.750 bao
MONEY_EXP = Decimal("0.01")       # 2 dp, VND

ZERO_QUANTITY = Decimal("0.000")
ZERO_MONEY = Decimal("0.00")


def _to_decimal(value: Decimal | int | float | str) -> Decimal:
    """Coerce to Decimal without inheriting float's binary error.

    `Decimal(0.1)` is 0.1000000000000000055511151231257827021181583404541015625.
    `Decimal(str(0.1))` is exactly 0.1. Always go through the string.
    """
    if isinstance(value, Decimal):
        return value
    try:
        return Decimal(str(value))
    except (InvalidOperation, ValueError, TypeError) as exc:
        raise ValueError(f"Không phải số hợp lệ: {value!r}") from exc


def quantize_quantity(value: Decimal | int | float | str) -> Decimal:
    """Round to 3 decimal places, half-up. Mirrors the client's rounding."""
    return _to_decimal(value).quantize(QUANTITY_EXP, rounding=ROUND_HALF_UP)


def quantize_money(value: Decimal | int | float | str) -> Decimal:
    """Round to 2 decimal places, half-up. VND."""
    return _to_decimal(value).quantize(MONEY_EXP, rounding=ROUND_HALF_UP)


def line_total(quantity: Decimal, unit_cost: Decimal) -> Decimal:
    """quantity x unit_cost, rounded to money precision.

    Rounded once, at the end, rather than at each step -- rounding the
    intermediate product would drift by a đồng per line and by a visible
    amount across a season's worth of transactions.
    """
    return quantize_money(quantize_quantity(quantity) * quantize_money(unit_cost))

File: mobile/src/utils/numeric.ts
/**
 * Money and quantity — the client half of the rounding contract.
 *
 * MUST agree with backend/app/core/numeric.py. Both sides round identically
 * at the boundary, so the stored value on device and server is bit-identical
 * and last-write-wins never fires on a value that merely *looks* different.
 *
 * See Data_Requirements_Database.md §7.1.
 */

/** Round to 3 decimal places — 0.250 kg, 12.500 L, 1.750 bao. */
export function quantizeQuantity(value: number): number {
  return Math.round(value * 1000) / 1000;
}

/** Round to 2 decimal places. VND. */
export function quantizeMoney(value: number): number {
  return Math.round(value * 100) / 100;
}

/**
 * quantity × unitCost, rounded once at the end.
 *
 * Rounding the intermediate product would drift by a đồng per line and by a
 * visible amount across a season's worth of transactions.
 */
export function lineTotal(quantity: number, unitCost: number): number {
  return quantizeMoney(quantizeQuantity(quantity) * quantizeMoney(unitCost));
}

/**
 * Parse what the farmer typed. Accepts '12,5' as well as '12.5' — Vietnamese
 * keyboards and habits produce the comma, and rejecting it would be a
 * pointless obstacle at the exact moment they are standing in a field.
 */
export function parseNumberInput(raw: string): number | null {
  const cleaned = raw.trim().replace(/\s/g, '').replace(',', '.');
  if (cleaned === '') {
    return null;
  }
  const value = Number(cleaned);
  return Number.isFinite(value) ? value : null;
}

// ─── Hiển thị ───────────────────────────────────────────────────────────────

const vndFormatter = new Intl.NumberFormat('vi-VN', {
  maximumFractionDigits: 0,
});

/** '1.250.000 ₫' */
export function formatMoney(value: number | null | undefined): string {
  if (value === null || value === undefined) {
    return '—';
  }
  return `${vndFormatter.format(Math.round(value))} ₫`;
}

/** '1,25 tr' / '1,2 tỷ' — for chart axes and tight tiles where the full
 *  number would not fit. */
export function formatMoneyShort(value: number | null | undefined): string {
  if (value === null || value === undefined) {
    return '—';
  }
  const abs = Math.abs(value);
  const sign = value < 0 ? '-' : '';
  if (abs >= 1_000_000_000) {
    return `${sign}${(abs / 1_000_000_000).toFixed(1).replace('.', ',')} tỷ`;
  }
  if (abs >= 1_000_000) {
    return `${sign}${(abs / 1_000_000).toFixed(1).replace('.', ',')} tr`;
  }
  if (abs >= 1_000) {
    return `${sign}${(abs / 1_000).toFixed(0)}k`;
  }
  return `${sign}${abs.toFixed(0)}`;
}

/** '12,5 kg' — trailing zeros trimmed, comma as the decimal separator. */
export function formatQuantity(value: number | null | undefined, unit?: string): string {
  if (value === null || value === undefined) {
    return '—';
  }
  const rounded = quantizeQuantity(value);
  const text = (Number.isInteger(rounded) ? String(rounded) : String(rounded)).replace(
    '.',
    ',',
  );
  return unit ? `${text} ${unit}` : text;
}
